API ML - Obtaining Information about API Services (#1520)
zowe/docs-site: docs/extend/extend-apiml/service-information.md

# Obtaining Information about API Services

## Overview

API clients can obtain obtain information about API services.

This is done for several use cases:

- Display available services to the user based on some criteria (API ID, hostname, or custom metadata)
- Locate a specific API service based on specific criteria (for example the API ID)
- Obtain information that allows routing via API gateway such as _baseUrl_ or _basePath_
- Obtain information about the API services, its APIs, or its service instances

Notes:

- _baseUrl_ or _basePath_ are explained in [Components of URL](api-mediation-components-of-URL.md)

> - The _API ID_ uniquely identifies the API in the API ML. The API ID can be used to locate the same APIs that are provided by different service instances. The API developer defines this ID. It must be a string of up to 64 characters that uses lowercase alphanumeric characters and a dot: `.`. The Zowe APIs start with the `zowe.` prefix. API ML APIs start with the `zowe.apiml.` prefix.

Reviewer: I am not sure whether details about API ID belongs here. The audience should be the client developer, not the API developer. It causes distraction in my opinion.

Author: You are right. I will remove the extra information.

## Protection of Service Information

The information about API services is categorized as sensitive. It provides partial information about the internal topology of the mainframe system and should be accessed only by authorized users and services.

Access to this information requires authentication using mainframe credentials and SAF resource check (SAF resource check - TBD).

## API Endpoints

### Obtain Information about Specific Service

`GET /gateway/api/v1/services/{serviceId}`

where:

- **{serviceId}** is the service ID of the API service (e.g. `apicatalog`)

It returns a JSON response that describes the service - see [Response Format](#response-format).

### Obtain Information about All Services

`GET /gateway/api/v1/services`

It returns a JSON response with a list of all services - see [Response Format](#response-format).

### Obtain Information about All Services with a Specific API ID

`GET /gateway/api/v1/services?apiId={apiId}`

where:

> - **{apiId}** is the API ID that needs to be provided by the service (e.g. `zowe.apiml.apicatalog`)

Reviewer: ```suggestion
- **{apiId}** is the API ID that represents required API (e.g. `zowe.apiml.apicatalog`)
```

Author: Thanks. That is better

## API ID in the API Mediation Layer

The _API ID_ uniquely identifies the API in the API ML. The API ID can be used to locate the same APIs that are provided by different service instances. The API developer defines this ID.

For more information about _baseUrl_ or _basePath_, see [Components of URL](api-mediation-components-of-URL.md).

## Protection of Service Information

Information about API services is considered sensitive as it contains partial information about the internal topology of the mainframe system. As such, this information should be made accessible only by authorized users and services.

Access to this information requires authentication using mainframe credentials, and a SAF resource check is done. The resource class and resource is defined in the `ZWESECUR` job. You can find more details about the `ZWESECUR` job in [Configuring the z/OS system for Zowe](../../user-guide/configure-zos-system.md).

The security administrator needs to permit READ access to the `APIML.SERVICES` resource in the `ZOWE` resource class to the your that can access the information about API services.

> 

Author: Add examples of PERMIT commands for ACF2, RACF, and TSS.
